{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Agent-based protein design, analysis and modeling "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ef7236db4790485d828cd9f2ccfdfdf4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using cached data from /tmp/chroma_weights/90e339502ae6b372797414167ce5a632/weights.pt\n",
      "Loaded from cache\n",
      "Using cached data from /tmp/chroma_weights/03a3a9af343ae74998768a2711c8b7ce/weights.pt\n",
      "Loaded from cache\n",
      "Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.\n",
      "Token is valid (permission: write).\n",
      "Your token has been saved to /home/alireza/.cache/huggingface/token\n",
      "Login successful\n"
     ]
    }
   ],
   "source": [
    "import autogen\n",
    "import os\n",
    "import tqdm \n",
    "import time\n",
    "import openai\n",
    "from openai import AsyncOpenAI\n",
    "from llama_index import VectorStoreIndex, SimpleDirectoryReader, ServiceContext\n",
    "from IPython.display import Markdown, display\n",
    "import py3Dmol\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from chroma import api\n",
    "from chroma import Chroma, Protein, conditioners\n",
    "from chroma.models import graph_classifier, procap\n",
    "from chroma.utility.api import register_key\n",
    "from chroma.utility.chroma import letter_to_point_cloud, plane_split_protein\n",
    "\n",
    "chroma_key = '###'\n",
    "HF_token= '###'\n",
    "OpenAI_key='sk-###'\n",
    "####################### CHROMA ##########################\n",
    "api.register_key(chroma_key,)\n",
    "chroma = Chroma()\n",
    "###################### HF\n",
    "from llama_index.embeddings import HuggingFaceEmbedding\n",
    "from llama_index.embeddings import OpenAIEmbedding\n",
    "\n",
    "token = HF_token\n",
    "from huggingface_hub import login\n",
    "login(token=token)\n",
    "\n",
    "#pip install transformers==4.35.1\n",
    "\n",
    "######### OPENAI ###########\n",
    "os.environ['OPENAI_API_KEY']=OpenAI_key"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using GPT-4 - Main setup starts here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('0.2.2', '1.6.0', '4.35.1')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "device = \"cpu\"\n",
    "import transformers\n",
    "autogen.__version__, openai.__version__, transformers.__version__ #make sure to update both autogen and openai to their newsest versions \n",
    "# ('0.2.2', '1.6.0', '4.35.1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from llm_config import llm_config\n",
    "from llm_config import config_list\n",
    "import agent_functions as func\n",
    "import agents\n",
    "import autogen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _reset_agents():\n",
    "    agents.user_proxy.reset()\n",
    "    agents.ragproxyagent.reset()\n",
    "    agents.planner.reset()\n",
    "    agents.assistant.reset()\n",
    "    agents.reviewer.reset()\n",
    "    agents.coder.reset()\n",
    "    agents.executor.reset()\n",
    "    #sequence_retriever.reset()\n",
    "#manager = autogen.GroupChatManager(groupchat=groupchat, llm_config=llm_config)\n",
    "_reset_agents()\n",
    "\n",
    "manager_llm_config = {\n",
    "    \"config_list\": config_list,\n",
    "    \"seed\": 45,\n",
    "}\n",
    "\n",
    "groupchat = autogen.GroupChat(\n",
    "    agents=[agents.user_proxy, agents.planner, agents.assistant, agents.critic, #agents.coder, agents.executor  #sequence_retriever,\n",
    "               ], messages=[], max_round=150, \n",
    ")\n",
    "manager = autogen.GroupChatManager(groupchat=groupchat, llm_config = manager_llm_config, \n",
    "                system_message='This agent repeats the following steps: Dynamically selecting a speaker, collecting response, and \\\n",
    "broadcasting the message to the group.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "user_proxy (to chat_manager):\n",
      "\n",
      "\n",
      "    Design three proteins with sequence length of 120 and analyze their secondary structure and first 10 frequencies. \n",
      "    Then fold the proteins and repeat the analysis. \n",
      "    Save all the results (including the AA sequences) into a csv file where each row corresponds to a protein and each column respresents a protein propery, utilizing proper identifiers.\n",
      "    \n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Planner (to chat_manager):\n",
      "\n",
      "Plan:\n",
      "\n",
      "1. Use the `design_protein_from_length` function three times to design three proteins each with a sequence length of 120. I will use 1, 2, 3 as names for the proteins. Save the amino acid sequences and PDB filenames.\n",
      "2. Use the `analyze_protein_structure` function to analyze the secondary structure of the proteins. Store the results.\n",
      "3. Use the `calc_protein_ANM` function to calculate the first 10 frequencies of the proteins. Store the results.\n",
      "4. Use the `fold_protein` function to fold the proteins. \n",
      "5. Repeat steps 2 and 3 with the folded proteins.\n",
      "6. Use the `save_to_csv_file` function to save all results into a csv file where each row corresponds to a protein and each column represents a protein property.\n",
      "7. The identifiers will be named as follows:\n",
      "  - Protein Name\n",
      "  - Amino Acid Sequence (Pre-Fold)\n",
      "  - Secondary Structure (Pre-Fold)\n",
      "  - Frequencies (Pre-Fold)\n",
      "  - Amino Acid Sequence (Post-Fold)\n",
      "  - Secondary Structure (Post-Fold)\n",
      "  - Frequencies (Post-Fold)\n",
      "\n",
      "Please confirm if this plan works for you.\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Critic (to chat_manager):\n",
      "\n",
      "Your plan is almost perfect. However, when a protein is folded, the protein sequence doesn't change, only the structure does, which then affects the secondary structure and possible frequencies you can calculate. So you need to only save Amino Acid Sequence once after designing the proteins. Each protein should have the following column properties:\n",
      "  \n",
      "  - Protein Number\n",
      "  - Amino Acid Sequence\n",
      "  - Secondary Structure (Pre-Fold)\n",
      "  - Frequencies (Pre-Fold)\n",
      "  - Secondary Structure (Post-Fold)\n",
      "  - Frequencies (Post-Fold)\n",
      "  \n",
      "Other than that, your plan works perfectly. So steps will be as follows:\n",
      "\n",
      "1. Use the `design_protein_from_length` function three times to design three proteins each with a sequence length of 120. I will use 1, 2, 3 as numbers for the proteins. Save the amino acid sequences and PDB filenames.\n",
      "2. Use the `analyze_protein_structure` function to analyze the secondary structure of the proteins. Store the results.\n",
      "3. Use the `calc_protein_ANM` function to calculate the first 10 frequencies of the proteins. Store the results.\n",
      "4. Use the `fold_protein` function to fold the proteins. Get the new structure names.\n",
      "5. Repeat steps 2 and 3 with the new folded protein structure.\n",
      "6. Use the `save_to_csv_file` function to save all results into a csv file where each row corresponds to a protein and each column represents a protein property.\n",
      "   \n",
      "Please note that in step 3, `calc_protein_ANM` will require `n_modes: 10` to calculate the first 10 frequencies.\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Provide feedback to chat_manager. Press enter to skip and use auto-reply, or type 'exit' to end the conversation: Excellent!\n",
      "user_proxy (to chat_manager):\n",
      "\n",
      "Excellent!\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: design_protein_from_length *****\n",
      "Arguments: \n",
      "{\n",
      "  \"length\": 120,\n",
      "  \"name\": 1\n",
      "}\n",
      "***************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION design_protein_from_length...\n",
      "Using cached data from /tmp/chroma_weights/90e339502ae6b372797414167ce5a632/weights.pt\n",
      "Loaded from cache\n",
      "Using cached data from /tmp/chroma_weights/03a3a9af343ae74998768a2711c8b7ce/weights.pt\n",
      "Loaded from cache\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8f61d2df5c549b98643707bce2809c3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Integrating SDE:   0%|          | 0/300 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e18ef066d16e4cba9eddbc2e02dfc736",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Potts Sampling:   0%|          | 0/500 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e33241abcc1c4f09a8a35006a0eba69a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Sequential decoding:   0%|          | 0/120 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"design_protein_from_length\" *****\n",
      "('./code_protein/1.pdb', 'MIIINIKTENGLSITYNSDEKKLELKYTPVKSPEDFKFPEDAKATISEVEYKGKKVIKIDAKLYVSPDLSKAKLTIEVNADISQEEADKIIDEFIKLLESLGNIKLKVTKDGNKYTIEVE')\n",
      "***********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: design_protein_from_length *****\n",
      "Arguments: \n",
      "{\n",
      "  \"length\": 120,\n",
      "  \"name\": 2\n",
      "}\n",
      "***************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION design_protein_from_length...\n",
      "Using cached data from /tmp/chroma_weights/90e339502ae6b372797414167ce5a632/weights.pt\n",
      "Loaded from cache\n",
      "Using cached data from /tmp/chroma_weights/03a3a9af343ae74998768a2711c8b7ce/weights.pt\n",
      "Loaded from cache\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2bf3d5fd80fa44d5adbaf97512f8f8ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Integrating SDE:   0%|          | 0/300 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "40e4a68c4ca64cccb3e12f9c48d757a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Potts Sampling:   0%|          | 0/500 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "11b327374fce404c8cf09eedbba826e3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Sequential decoding:   0%|          | 0/120 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"design_protein_from_length\" *****\n",
      "('./code_protein/2.pdb', 'GSPLPRPPLSPEEQEALRKKAQEKYNEFVSKIKELLRRAADRVRRGEPVELIEKTIKIGDYEYKIVATSPEEAKELENLIKEMIDLGFKPSKEFSDKLVEAARLIREGRVDEALRLLDEM')\n",
      "***********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: design_protein_from_length *****\n",
      "Arguments: \n",
      "{\n",
      "  \"length\": 120,\n",
      "  \"name\": 3\n",
      "}\n",
      "***************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION design_protein_from_length...\n",
      "Using cached data from /tmp/chroma_weights/90e339502ae6b372797414167ce5a632/weights.pt\n",
      "Loaded from cache\n",
      "Using cached data from /tmp/chroma_weights/03a3a9af343ae74998768a2711c8b7ce/weights.pt\n",
      "Loaded from cache\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5fb9fe81c9dd4013a471a20d30683cc0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Integrating SDE:   0%|          | 0/300 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c40503413f404360a6465c2e9f59c932",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Potts Sampling:   0%|          | 0/500 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "86042e8dbce74c8f98045693c033d71b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Sequential decoding:   0%|          | 0/120 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"design_protein_from_length\" *****\n",
      "('./code_protein/3.pdb', 'APLDPDDLSAQLRAAIDELVRLGYEEEVSKPEFIEALRLYALDLGLKEVVLRRVTPAPASQPGVYTVEDVTVDLEALRKQELSPEEQARLEKIRAKYDEMLADPEFQALLDEVLARARAA')\n",
      "***********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Critic (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/1.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: ./code_protein/1.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 13.333333333333334,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 46.666666666666664,\n",
      "    \"G\": 0.0,\n",
      "    \"I\": 0.0,\n",
      "    \"T\": 14.166666666666666,\n",
      "    \"S\": 7.5,\n",
      "    \"P\": 0.0,\n",
      "    \"-\": 18.333333333333332\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/2.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: ./code_protein/2.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 61.66666666666667,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 11.666666666666666,\n",
      "    \"G\": 0.0,\n",
      "    \"I\": 0.0,\n",
      "    \"T\": 7.5,\n",
      "    \"S\": 3.3333333333333335,\n",
      "    \"P\": 3.3333333333333335,\n",
      "    \"-\": 12.5\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/3.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: ./code_protein/3.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 57.49999999999999,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 13.333333333333334,\n",
      "    \"G\": 0.0,\n",
      "    \"I\": 4.166666666666666,\n",
      "    \"T\": 8.333333333333332,\n",
      "    \"S\": 3.3333333333333335,\n",
      "    \"P\": 6.666666666666667,\n",
      "    \"-\": 6.666666666666667\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/1.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 956 atoms and 1 coordinate set(s) were parsed in 0.01s.\n",
      "DEBUG:.prody:956 atoms and 1 coordinate set(s) were parsed in 0.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: ./code_protein/1.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.90s.\n",
      "DEBUG:.prody:Hessian was built in 0.90s.\n",
      "@> 10 modes were calculated in 1.10s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.10s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    2.0337,\n",
      "    2.8678,\n",
      "    3.3843,\n",
      "    3.6263,\n",
      "    3.9904,\n",
      "    4.5381,\n",
      "    4.8373,\n",
      "    4.8956,\n",
      "    5.1492,\n",
      "    5.4416\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/2.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 974 atoms and 1 coordinate set(s) were parsed in 0.01s.\n",
      "DEBUG:.prody:974 atoms and 1 coordinate set(s) were parsed in 0.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: ./code_protein/2.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.85s.\n",
      "DEBUG:.prody:Hessian was built in 0.85s.\n",
      "@> 10 modes were calculated in 1.15s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.15s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    0.0207,\n",
      "    0.1058,\n",
      "    0.1782,\n",
      "    0.4189,\n",
      "    0.49,\n",
      "    0.9015,\n",
      "    1.1832,\n",
      "    1.8257,\n",
      "    2.1212,\n",
      "    2.8726\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"./code_protein/3.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 946 atoms and 1 coordinate set(s) were parsed in 0.01s.\n",
      "DEBUG:.prody:946 atoms and 1 coordinate set(s) were parsed in 0.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: ./code_protein/3.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.80s.\n",
      "DEBUG:.prody:Hessian was built in 0.80s.\n",
      "@> 10 modes were calculated in 1.07s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.07s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    0.7546,\n",
      "    1.0836,\n",
      "    1.5026,\n",
      "    1.8874,\n",
      "    2.0844,\n",
      "    2.3192,\n",
      "    2.7975,\n",
      "    3.0199,\n",
      "    3.0669,\n",
      "    3.1382\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Planner (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: fold_protein *****\n",
      "Arguments: \n",
      "{\n",
      "  \"sequence\": \"MIIINIKTENGLSITYNSDEKKLELKYTPVKSPEDFKFPEDAKATISEVEYKGKKVIKIDAKLYVSPDLSKAKLTIEVNADISQEEADKIIDEFIKLLESLGNIKLKVTKDGNKYTIEVE\",\n",
      "  \"name\": \"1_folded\"\n",
      "}\n",
      "*************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION fold_protein...\n",
      "INFO:root:Loading weights from /home/alireza/.cache/omegafold_ckpt/model2.pt\n",
      "INFO:root:Constructing OmegaFold\n",
      "INFO:root:Reading temp.fasta\n",
      "INFO:root:Predicting 1th chain in temp.fasta\n",
      "INFO:root:120 residues in this chain.\n",
      "INFO:root:Finished prediction in 243.32 seconds.\n",
      "INFO:root:Saving prediction to ./code_protein/1_folded.pdb\n",
      "INFO:root:Saved\n",
      "INFO:root:Done!\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"fold_protein\" *****\n",
      "1_folded.pdb\n",
      "*********************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: fold_protein *****\n",
      "Arguments: \n",
      "{\n",
      "  \"sequence\": \"GSPLPRPPLSPEEQEALRKKAQEKYNEFVSKIKELLRRAADRVRRGEPVELIEKTIKIGDYEYKIVATSPEEAKELENLIKEMIDLGFKPSKEFSDKLVEAARLIREGRVDEALRLLDEM\",\n",
      "  \"name\": \"2_folded\"\n",
      "}\n",
      "*************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION fold_protein...\n",
      "INFO:root:Loading weights from /home/alireza/.cache/omegafold_ckpt/model2.pt\n",
      "INFO:root:Constructing OmegaFold\n",
      "INFO:root:Reading temp.fasta\n",
      "INFO:root:Predicting 1th chain in temp.fasta\n",
      "INFO:root:120 residues in this chain.\n",
      "INFO:root:Finished prediction in 247.16 seconds.\n",
      "INFO:root:Saving prediction to ./code_protein/2_folded.pdb\n",
      "INFO:root:Saved\n",
      "INFO:root:Done!\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"fold_protein\" *****\n",
      "2_folded.pdb\n",
      "*********************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: fold_protein *****\n",
      "Arguments: \n",
      "{\n",
      "  \"sequence\": \"APLDPDDLSAQLRAAIDELVRLGYEEEVSKPEFIEALRLYALDLGLKEVVLRRVTPAPASQPGVYTVEDVTVDLEALRKQELSPEEQARLEKIRAKYDEMLADPEFQALLDEVLARARAA\",\n",
      "  \"name\": \"3_folded\"\n",
      "}\n",
      "*************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION fold_protein...\n",
      "INFO:root:Loading weights from /home/alireza/.cache/omegafold_ckpt/model2.pt\n",
      "INFO:root:Constructing OmegaFold\n",
      "INFO:root:Reading temp.fasta\n",
      "INFO:root:Predicting 1th chain in temp.fasta\n",
      "INFO:root:120 residues in this chain.\n",
      "INFO:root:Finished prediction in 267.79 seconds.\n",
      "INFO:root:Saving prediction to ./code_protein/3_folded.pdb\n",
      "INFO:root:Saved\n",
      "INFO:root:Done!\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"fold_protein\" *****\n",
      "3_folded.pdb\n",
      "*********************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"1_folded.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: code_protein/1_folded.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 15.833333333333332,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 46.666666666666664,\n",
      "    \"G\": 2.5,\n",
      "    \"I\": 0.0,\n",
      "    \"T\": 14.166666666666666,\n",
      "    \"S\": 4.166666666666666,\n",
      "    \"P\": 0.0,\n",
      "    \"-\": 16.666666666666664\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"2_folded.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: code_protein/2_folded.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 62.5,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 11.666666666666666,\n",
      "    \"G\": 0.0,\n",
      "    \"I\": 0.0,\n",
      "    \"T\": 6.666666666666667,\n",
      "    \"S\": 1.6666666666666667,\n",
      "    \"P\": 4.166666666666666,\n",
      "    \"-\": 13.333333333333334\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: analyze_protein_structure *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"3_folded.pdb\"\n",
      "}\n",
      "**************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION analyze_protein_structure...\n",
      "Analyzing the secondary strucute of this protein: code_protein/3_folded.pdb\n",
      "The protein analyzed has  120 residues.\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"analyze_protein_structure\" *****\n",
      "{\n",
      "    \"H\": 61.66666666666667,\n",
      "    \"B\": 0.0,\n",
      "    \"E\": 15.0,\n",
      "    \"G\": 0.0,\n",
      "    \"I\": 0.0,\n",
      "    \"T\": 8.333333333333332,\n",
      "    \"S\": 3.3333333333333335,\n",
      "    \"P\": 1.6666666666666667,\n",
      "    \"-\": 10.0\n",
      "}\n",
      "**********************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"1_folded.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 956 atoms and 1 coordinate set(s) were parsed in 0.03s.\n",
      "DEBUG:.prody:956 atoms and 1 coordinate set(s) were parsed in 0.03s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: code_protein/1_folded.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.88s.\n",
      "DEBUG:.prody:Hessian was built in 0.88s.\n",
      "@> 10 modes were calculated in 1.10s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.10s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    1.8739,\n",
      "    2.1563,\n",
      "    2.7611,\n",
      "    3.1086,\n",
      "    3.8712,\n",
      "    4.0481,\n",
      "    4.3759,\n",
      "    4.6717,\n",
      "    4.8183,\n",
      "    4.9126\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"2_folded.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 974 atoms and 1 coordinate set(s) were parsed in 0.01s.\n",
      "DEBUG:.prody:974 atoms and 1 coordinate set(s) were parsed in 0.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: code_protein/2_folded.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.88s.\n",
      "DEBUG:.prody:Hessian was built in 0.88s.\n",
      "@> 10 modes were calculated in 1.11s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.11s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    0.0444,\n",
      "    0.1641,\n",
      "    0.3379,\n",
      "    0.5724,\n",
      "    0.765,\n",
      "    0.9568,\n",
      "    1.4306,\n",
      "    1.5344,\n",
      "    1.6834,\n",
      "    1.8099\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: calc_protein_ANM *****\n",
      "Arguments: \n",
      "{\n",
      "  \"protein_structure\": \"3_folded.pdb\",\n",
      "  \"n_modes\": 10\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION calc_protein_ANM...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 946 atoms and 1 coordinate set(s) were parsed in 0.01s.\n",
      "DEBUG:.prody:946 atoms and 1 coordinate set(s) were parsed in 0.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing ANM for protein structure: code_protein/3_folded.pdb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> Hessian was built in 0.75s.\n",
      "DEBUG:.prody:Hessian was built in 0.75s.\n",
      "@> 10 modes were calculated in 1.01s.\n",
      "DEBUG:.prody:10 modes were calculated in 1.01s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"calc_protein_ANM\" *****\n",
      "[\n",
      "    0.5256,\n",
      "    1.0278,\n",
      "    1.1566,\n",
      "    1.2877,\n",
      "    1.5521,\n",
      "    1.9111,\n",
      "    2.1887,\n",
      "    2.4664,\n",
      "    2.734,\n",
      "    2.8731\n",
      "]\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Planner (to chat_manager):\n",
      "\n",
      "***** Suggested function Call: save_to_csv_file *****\n",
      "Arguments: \n",
      "{\n",
      "  \"input_JSON_dictionary\": \"{\n",
      "    'Protein Number': ['1', '2', '3'],\n",
      "    'Amino Acid Sequence': [\n",
      "      'MIIINIKTENGLSITYNSDEKKLELKYTPVKSPEDFKFPEDAKATISEVEYKGKKVIKIDAKLYVSPDLSKAKLTIEVNADISQEEADKIIDEFIKLLESLGNIKLKVTKDGNKYTIEVE',\n",
      "      'GSPLPRPPLSPEEQEALRKKAQEKYNEFVSKIKELLRRAADRVRRGEPVELIEKTIKIGDYEYKIVATSPEEAKELENLIKEMIDLGFKPSKEFSDKLVEAARLIREGRVDEALRLLDEM',\n",
      "      'APLDPDDLSAQLRAAIDELVRLGYEEEVSKPEFIEALRLYALDLGLKEVVLRRVTPAPASQPGVYTVEDVTVDLEALRKQELSPEEQARLEKIRAKYDEMLADPEFQALLDEVLARARAA'\n",
      "    ],\n",
      "    'Secondary Structure (Pre-Fold)': [\n",
      "      {'H': 13.333333333333334, 'B': 0.0, 'E': 46.666666666666664, 'G': 0.0, 'I': 0.0, 'T': 14.166666666666666, 'S': 7.5, 'P': 0.0, '-': 18.333333333333332},\n",
      "      {'H': 61.66666666666667, 'B': 0.0, 'E': 11.666666666666666, 'G': 0.0, 'I': 0.0, 'T': 7.5, 'S': 3.3333333333333335, 'P': 3.3333333333333335, '-': 12.5},\n",
      "      {'H': 57.49999999999999, 'B': 0.0, 'E': 13.333333333333334, 'G': 0.0, 'I': 4.166666666666666, 'T': 8.333333333333332, 'S': 3.3333333333333335, 'P': 6.666666666666667, '-': 6.666666666666667}\n",
      "    ],\n",
      "    'Frequencies (Pre-Fold)': [\n",
      "      [2.0337, 2.8678, 3.3843, 3.6263, 3.9904, 4.5381, 4.8373, 4.8956, 5.1492, 5.4416],\n",
      "      [0.0207, 0.1058, 0.1782, 0.4189, 0.49, 0.9015, 1.1832, 1.8257, 2.1212, 2.8726],\n",
      "      [0.7546, 1.0836, 1.5026, 1.8874, 2.0844, 2.3192, 2.7975, 3.0199, 3.0669, 3.1382]\n",
      "    ],\n",
      "    'Secondary Structure (Post-Fold)': [\n",
      "      {'H': 15.833333333333332, 'B': 0.0, 'E': 46.666666666666664, 'G': 2.5, 'I': 0.0, 'T': 14.166666666666666, 'S': 4.166666666666666, 'P': 0.0, '-': 16.666666666666664},\n",
      "      {'H': 62.5, 'B': 0.0, 'E': 11.666666666666666, 'G': 0.0, 'I': 0.0, 'T': 6.666666666666667, 'S': 1.6666666666666667, 'P': 4.166666666666666, '-': 13.333333333333334},\n",
      "      {'H': 61.66666666666667, 'B': 0.0, 'E': 15.0, 'G': 0.0, 'I': 0.0, 'T': 8.333333333333332, 'S': 3.3333333333333335, 'P': 1.6666666666666667, '-': 10.0}\n",
      "    ],\n",
      "    'Frequencies (Post-Fold)': [\n",
      "      [1.8739, 2.1563, 2.7611, 3.1086, 3.8712, 4.0481, 4.3759, 4.6717, 4.8183, 4.9126],\n",
      "      [0.0444, 0.1641, 0.3379, 0.5724, 0.765, 0.9568, 1.4306, 1.5344, 1.6834, 1.8099],\n",
      "      [0.5256, 1.0278, 1.1566, 1.2877, 1.5521, 1.9111, 2.1887, 2.4664, 2.734, 2.8731]\n",
      "    ]\n",
      "  }\",\n",
      "  \"output_csv_name\": \"protein_data.csv\"\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION save_to_csv_file...\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"save_to_csv_file\" *****\n",
      "Error: Expecting property name enclosed in double quotes: line 2 column 5 (char 6)\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Critic (to chat_manager):\n",
      "\n",
      "I apologize for the mistake. It seems there was an issue with the formatting of the input JSON dictionary for saving to the CSV file. Let's try again with properly formatted JSON.\n",
      "\n",
      "Plan to fix the error:\n",
      "\n",
      "1. Correctly format JSON with double quotes instead of single quotations.\n",
      "2. Call 'save_to_csv_file' function again with correctly formatted JSON dictionary.\n",
      "\n",
      "\n",
      "I'll directly proceed with this plan as the error was due to syntax not plan. \n",
      "Here is the corrected data:\n",
      "***** Suggested function Call: save_to_csv_file *****\n",
      "Arguments: \n",
      "{\n",
      "  \"input_JSON_dictionary\": \"{\\\"Protein Number\\\": [\\\"1\\\", \\\"2\\\", \\\"3\\\"],\\\"Amino Acid Sequence\\\": [\\\"MIIINIKTENGLSITYNSDEKKLELKYTPVKSPEDFKFPEDAKATISEVEYKGKKVIKIDAKLYVSPDLSKAKLTIEVNADISQEEADKIIDEFIKLLESLGNIKLKVTKDGNKYTIEVE\\\",\\\"GSPLPRPPLSPEEQEALRKKAQEKYNEFVSKIKELLRRAADRVRRGEPVELIEKTIKIGDYEYKIVATSPEEAKELENLIKEMIDLGFKPSKEFSDKLVEAARLIREGRVDEALRLLDEM\\\",\\\"APLDPDDLSAQLRAAIDELVRLGYEEEVSKPEFIEALRLYALDLGLKEVVLRRVTPAPASQPGVYTVEDVTVDLEALRKQELSPEEQARLEKIRAKYDEMLADPEFQALLDEVLARARAA\\\"],\\\"Secondary Structure (Pre-Fold)\\\": [{\\\"H\\\": 13.333333333333334, \\\"B\\\": 0.0, \\\"E\\\": 46.666666666666664, \\\"G\\\": 0.0, \\\"I\\\": 0.0, \\\"T\\\": 14.166666666666666, \\\"S\\\": 7.5, \\\"P\\\": 0.0, \\\"-\\\": 18.333333333333332},{\\\"H\\\": 61.66666666666667, \\\"B\\\": 0.0, \\\"E\\\": 11.666666666666666, \\\"G\\\": 0.0, \\\"I\\\": 0.0, \\\"T\\\": 7.5, \\\"S\\\": 3.3333333333333335, \\\"P\\\": 3.3333333333333335, \\\"-\\\": 12.5},{\\\"H\\\": 57.49999999999999, \\\"B\\\": 0.0, \\\"E\\\": 13.333333333333334, \\\"G\\\": 0.0, \\\"I\\\": 4.166666666666666, \\\"T\\\": 8.333333333333332, \\\"S\\\": 3.3333333333333335, \\\"P\\\": 6.666666666666667, \\\"-\\\": 6.666666666666667}],\\\"Frequencies (Pre-Fold)\\\": [[2.0337, 2.8678, 3.3843, 3.6263, 3.9904, 4.5381, 4.8373, 4.8956, 5.1492, 5.4416],[0.0207, 0.1058, 0.1782, 0.4189, 0.49, 0.9015, 1.1832, 1.8257, 2.1212, 2.8726],[0.7546, 1.0836, 1.5026, 1.8874, 2.0844, 2.3192, 2.7975, 3.0199, 3.0669, 3.1382]],\\\"Secondary Structure (Post-Fold)\\\": [{\\\"H\\\": 15.833333333333332, \\\"B\\\": 0.0, \\\"E\\\": 46.666666666666664, \\\"G\\\": 2.5, \\\"I\\\": 0.0, \\\"T\\\": 14.166666666666666, \\\"S\\\": 4.166666666666666, \\\"P\\\": 0.0, \\\"-\\\": 16.666666666666664},{\\\"H\\\": 62.5, \\\"B\\\": 0.0, \\\"E\\\": 11.666666666666666, \\\"G\\\": 0.0, \\\"I\\\": 0.0, \\\"T\\\": 6.666666666666667, \\\"S\\\": 1.6666666666666667, \\\"P\\\": 4.166666666666666, \\\"-\\\": 13.333333333333334},{\\\"H\\\": 61.66666666666667, \\\"B\\\": 0.0, \\\"E\\\": 15.0, \\\"G\\\": 0.0, \\\"I\\\": 0.0, \\\"T\\\": 8.333333333333332, \\\"S\\\": 3.3333333333333335, \\\"P\\\": 1.6666666666666667, \\\"-\\\": 10.0}],\\\"Frequencies (Post-Fold)\\\": [[1.8739, 2.1563, 2.7611, 3.1086, 3.8712, 4.0481, 4.3759, 4.6717, 4.8183, 4.9126],[0.0444, 0.1641, 0.3379, 0.5724, 0.765, 0.9568, 1.4306, 1.5344, 1.6834, 1.8099],[0.5256, 1.0278, 1.1566, 1.2877, 1.5521, 1.9111, 2.1887, 2.4664, 2.734, 2.8731]]}\",\n",
      "  \"output_csv_name\": \"protein_data.csv\"\n",
      "}\n",
      "*****************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      ">>>>>>>> EXECUTING FUNCTION save_to_csv_file...\n",
      "the results have been saved to csv file: protein_data.csv\n",
      "assistant (to chat_manager):\n",
      "\n",
      "***** Response from calling function \"save_to_csv_file\" *****\n",
      "None\n",
      "*************************************************************\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:autogen.agentchat.groupchat:GroupChat select_speaker failed to resolve the next speaker's name. This is because the speaker selection OAI call returned:\n",
      "TERMINATE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Critic (to chat_manager):\n",
      "\n",
      "The plan was executed successfully and all data including Amino Acid Sequence, Secondary Structure (Pre-Fold and Post-Fold), and Frequencies (Pre-Fold and Post-Fold) for three proteins have been saved into a CSV file named 'protein_data.csv'. Each row of the file corresponds to a protein and each column represents a protein property. Let me know if you need any further assistance.\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:autogen.agentchat.groupchat:GroupChat select_speaker failed to resolve the next speaker's name. This is because the speaker selection OAI call returned:\n",
      "TERMINATE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Provide feedback to chat_manager. Press enter to skip and use auto-reply, or type 'exit' to end the conversation: Good job, Thanks a lot!\n",
      "user_proxy (to chat_manager):\n",
      "\n",
      "Good job, Thanks a lot!\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:autogen.agentchat.groupchat:GroupChat select_speaker failed to resolve the next speaker's name. This is because the speaker selection OAI call returned:\n",
      "TERMINATE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Planner (to chat_manager):\n",
      "\n",
      "You're welcome! I'm glad I could assist. If you have any other tasks or need further assistance, feel free to ask. Have a great day!\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:autogen.agentchat.groupchat:GroupChat select_speaker failed to resolve the next speaker's name. This is because the speaker selection OAI call returned:\n",
      "TERMINATE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assistant (to chat_manager):\n",
      "\n",
      "TERMINATE\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "agents.user_proxy.initiate_chat(\n",
    "    recipient=manager,\n",
    "    message='''\n",
    "    Design three proteins with sequence length of 120 and analyze their secondary structure and first 10 frequencies. \n",
    "    Then fold the proteins and repeat the analysis. \n",
    "    Save all the results (including the AA sequences) into a csv file where each row corresponds to a protein and each column respresents a protein propery, utilizing proper identifiers.\n",
    "    '''\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
